ext/opentelemetry-ext-pyramid/src/opentelemetry/ext/pyramid/callbacks.py

```diff
@@ -0,0 +1,172 @@
+
+_ENVIRON_STARTTIME_KEY = "opentelemetry-pyramid.starttime_key"
+_ENVIRON_SPAN_KEY = "opentelemetry-pyramid.span_key"
+_ENVIRON_ACTIVATION_KEY = "opentelemetry-pyramid.activation_key"
+_ENVIRON_ENABLED_KEY = "opentelemetry-pyramid.tracing_enabled_key"
+_ENVIRON_TOKEN = "opentelemetry-pyramid.token"
+
+_logger = getLogger(__name__)
+
+
+def get_excluded_hosts():
+    hosts = configuration.Configuration().PYRAMID_EXCLUDED_HOSTS or []
+    if hosts:
+        hosts = str.split(hosts, ",")
+    return hosts
+
+
+def get_excluded_paths():
+    paths = configuration.Configuration().PYRAMID_EXCLUDED_PATHS or []
+    if paths:
+        paths = str.split(paths, ",")
+    return paths
+
+
+_excluded_hosts = get_excluded_hosts()
+_excluded_paths = get_excluded_paths()
+
+
+def includeme(config):
+    config.add_settings({SETTING_TRACE_ENABLED: True})
+
+    config.add_subscriber(_before_traversal, BeforeTraversal)
+    _insert_tween(config)
+
+
+def _insert_tween(config):
+    settings = config.get_settings()
+    tweens = settings.get("pyramid.tweens")
+    # If the list is empty, pyramid does not consider the tweens have been
+    # set explicitly. And if our tween is already there, nothing to do
+    if not tweens or not tweens.strip():
+        # Add our tween just before the default exception handler
+        config.add_tween(TWEEN_NAME, over=EXCVIEW)
+
+
+def _before_traversal(event):
+    request = event.request
+    environ = request.environ
+    span_name = otel_wsgi.get_default_span_name(environ)
+
+    enabled = environ.get(_ENVIRON_ENABLED_KEY)
+    if enabled is None:
+        _logger.warning(
+            "Opentelemetry pyramid tween 'opentelemetry.ext.pyramid.trace_tween_factory'"
+            "was not called. Make sure that the tween is included in 'pyramid.tweens' if"
+            "the tween list was created manually"
+        )
+        return
+
+    if not enabled:
+        # Tracing not enabled, return
+        return
+
+    start_time = environ.get(_ENVIRON_STARTTIME_KEY)
+
+    token = context.attach(
+        propagators.extract(otel_wsgi.get_header_from_environ, environ)
+    )
+    tracer = trace.get_tracer(__name__, __version__)
+
+    attributes = otel_wsgi.collect_request_attributes(environ)
+
+    if request.matched_route:
+        span_name = request.matched_route.pattern
+        attributes["http.route"] = request.matched_route.pattern
+    else:
+        span_name = otel_wsgi.get_default_span_name(environ)
+
+    span = tracer.start_span(
+        span_name,
+        kind=trace.SpanKind.SERVER,
+        attributes=attributes,
+        start_time=start_time,
+    )
+
+    activation = tracer.use_span(span, end_on_exit=True)
+    activation.__enter__()
+    environ[_ENVIRON_ACTIVATION_KEY] = activation
+    environ[_ENVIRON_SPAN_KEY] = span
+    environ[_ENVIRON_TOKEN] = token
+
+
+def trace_tween_factory(handler, registry):
+    settings = registry.settings
+    enabled = asbool(settings.get(SETTING_TRACE_ENABLED, True))
```

Comment:
As I understand this function, if tracing is disabled during the initialization for the application, then we will never instrument the application at any point for its runtime, correct? as the trace_tween_factory is only called once during initialization, and it in the case where the trace is disabled, it'll return the disabled tween.

IOW, modifying the settings afterward has no impact on what is being traced.

Reply:
I haven't actually verified this, but I'm pretty sure the tween factory is called whenever `make_wsgi_app` is called. so if we `uninstrument_config()` which sets `enabled = False`, and then export the wsgi app, enabled would end up being False. If this wasn't the case I'm not sure how [this test](https://github.com/open-telemetry/opentelemetry-python/pull/776/files/0187389ae242a07047f4003f1d1a58a51fd80903#diff-4324378011de9fbc733916ad846df28bR61) would work, since we never removed the tween